### PR TITLE
Add option in ros2 topic pub to wait for N matching subscriptions, use N=1 by default when combined with --times

### PR DESCRIPTION
**Pull request by ivanpauno**
Replaces https://github.com/ros2/ros2cli/pull/641, seems to be a better idea to me.

### CODE REVIEW

**Review by audrow — APPROVED** [review]
Looks good to me with green CI.

Also, it would be great to add a test for this.

**Review by hidmic — APPROVED** [review]
LGTM, +1 to that test.

**Review by jacobperron — COMMENTED** [review]
LGTM with a test.

**Review by fujitatomoya — APPROVED** [review]

File: ros2topic/ros2topic/verb/pub.py
Comment (clalancette) [inline ros2topic/ros2topic/verb/pub.py:84]: I'll still argue that having the default be `1` is a substantial difference in behavior from earlier versions of ROS 2 and ROS 1.  As long as we are OK with that difference in behavior, then we can go forward, but I also think we then need to have a release note about it.

File: ros2topic/ros2topic/verb/pub.py
Comment (clalancette) [inline ros2topic/ros2topic/verb/pub.py:184]: I'll suggest:

```suggestion
            print(f'Waiting for at least {wait_matching_subscriptions} matching subscription(s)...')
```

File: ros2topic/ros2topic/verb/pub.py
Comment (ivanpauno) [reply ros2topic/ros2topic/verb/pub.py:84]: > As long as we are OK with that difference in behavior, then we can go forward, but I also think we then need to have a release note about it.

I can bring this up in the next ROS 2 meeting and ask for opinions, mentioning this in the release notes sounds fine to me.

> I'll still argue that having the default be 1 is a substantial difference in behavior from earlier versions of ROS 2 and ROS 1

This wasn't a problem in ROS 1 as it has an unique graph state that all nodes are aware.
When you created a node and a publisher, it was guaranteed that you were going to see all already existing subscriptions.

This is not the case in ROS 2, and I have got confused with `ros2 topic pub -1 --qos-reliability reliable <topic_name> <topic_type> <message>` not working enough times to think the default behavior isn't intuitive.

Ideally, we would also have https://github.com/ros2/rmw/issues/295 and call it before ending the process, specially when using `--times/--once`.

File: ros2topic/ros2topic/verb/pub.py
Comment (ivanpauno) [reply ros2topic/ros2topic/verb/pub.py:184]:
```suggestion
            print(
                f'Waiting for at least {wait_matching_subscriptions} matching subscription(s)...')
```

Comment (ivanpauno) [reply ros2topic/ros2topic/verb/pub.py]: Fixed in https://github.com/ros2/ros2cli/pull/642/commits/3d59b36971615b8cab6f733578023b182b833864

**Comment by ivanpauno** [comment]
@ros-pull-request-builder retest this please

**Comment by ivanpauno** [comment]
I've added a test in https://github.com/ros2/ros2cli/pull/642/commits/19ce63631a8c8c47cdd74175ccb05ff9ccab3211.

File: ros2topic/test/test_echo_pub.py
Comment (hidmic) [inline ros2topic/test/test_echo_pub.py:161]: @ivanpauno why the braces? Those go away after `f-string` interpolation.

Comment (hidmic) [reply ros2topic/test/test_echo_pub.py:161]: Also why `[0-7]+`? Wouldn't `[0-9]+` be more comprehensive?

**Review by hidmic — APPROVED** [review]

File: ros2topic/test/test_echo_pub.py
Comment (ivanpauno) [reply ros2topic/test/test_echo_pub.py:161]: > Also why [0-7]+? Wouldn't [0-9]+ be more comprehensive?

lol
I've to double-check this, I'm not sure if this filter is doing anything at all.

Comment (ivanpauno) [reply ros2topic/test/test_echo_pub.py]: I have delted the filter, as we aren't asserting that the output of the command matches something.

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14706/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9458/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12371/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14850/)

File: ros2topic/ros2topic/verb/pub.py
Comment (clalancette) [inline ros2topic/ros2topic/verb/pub.py:87]: This is very subjective, but `-w` feels more natural to me as the short version of the flag.  What do others think?

Comment (ivanpauno) [reply ros2topic/ros2topic/verb/pub.py:87]: Sounds fair to me.

I think I picked `-m` because I also was going to rename the long flag to `--matching-subscriptions`, but I ended up mixing both.
I don't have a strong opinion, either is fine to me.

Comment (ivanpauno) [reply ros2topic/ros2topic/verb/pub.py]: Fixed in https://github.com/ros2/ros2cli/pull/642/commits/bf92afecf75376ff2d08c1b39e802d875dafb93c.

**Comment by ivanpauno** [comment]
I have changed the default value according to the offline discussion in https://github.com/ros2/ros2cli/pull/642/commits/93cf8474fb4a4334a8438530adf03897cb803cb1.

**Review by clalancette — APPROVED** [review]
This all seems reasonable to me now, so I'm approving.  A few things before we merge this:

1. Green CI.
2. I think a release note in https://github.com/ros2/ros2_documentation/blob/rolling/source/Releases/Release-Humble-Hawksbill.rst is appropriate.
3. I think we should get one more approval before merging this.

**Review by hidmic — APPROVED** [review]

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14718/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9470/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12385/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14864/)

**Comment by ivanpauno** [comment]
CI is green, there're two approvals, PR to document this in the `Humble` release notes is on the way https://github.com/ros2/ros2_documentation/pull/1654.
Going in!!

**Pull request merged**
